src/Stratis.Bitcoin.Features.SmartContracts/Consensus/Rules/SmartContractPosCoinviewRule.cs

```diff
@@ -0,0 +1,367 @@
+        {
+            this.Logger.LogTrace("({0}:{1},{2}:'{3}')", nameof(fees), fees, nameof(height), height);
+
+            if (BlockStake.IsProofOfStake(block))
+            {
+                var posRuleContext = context as PosRuleContext;
+                Money stakeReward = block.Transactions[1].TotalOut - posRuleContext.TotalCoinStakeValueIn;
+                Money calcStakeReward = fees + this.GetProofOfStakeReward(height);
+
+                this.Logger.LogTrace("Block stake reward is {0}, calculated reward is {1}.", stakeReward, calcStakeReward);
+                if (stakeReward > calcStakeReward)
+                {
+                    this.Logger.LogTrace("(-)[BAD_COINSTAKE_AMOUNT]");
+                    ConsensusErrors.BadCoinstakeAmount.Throw();
+                }
+            }
+            else
+            {
+                Money blockReward = fees + this.GetProofOfWorkReward(height);
+                this.Logger.LogTrace("Block reward is {0}, calculated reward is {1}.", block.Transactions[0].TotalOut, blockReward);
+                if (block.Transactions[0].TotalOut > blockReward)
+                {
+                    this.Logger.LogTrace("(-)[BAD_COINBASE_AMOUNT]");
+                    ConsensusErrors.BadCoinbaseAmount.Throw();
+                }
+            }
+
+            this.Logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc />
+        public override void UpdateCoinView(RuleContext context, Transaction transaction)
+        {
+            this.Logger.LogTrace("()");
+
+            if (this.generatedTransaction != null)
+            {
+                ValidateGeneratedTransaction(transaction);
+                base.UpdateUTXOSet(context, transaction);
+                return;
+            }
+
+            // If we are here, was definitely submitted by someone
+            ValidateSubmittedTransaction(transaction);
+
+            TxOut smartContractTxOut = transaction.Outputs.FirstOrDefault(txOut => txOut.ScriptPubKey.IsSmartContractExec());
+            if (smartContractTxOut == null)
+            {
+                // Someone submitted a standard transaction - no smart contract opcodes.
+                var posRuleContext = context as PosRuleContext;
+                UnspentOutputSet view = posRuleContext.UnspentOutputSet;
+                if (transaction.IsCoinStake)
+                    posRuleContext.TotalCoinStakeValueIn = view.GetValueIn(transaction);
+
+                base.UpdateUTXOSet(context, transaction);
+
+                return;
+            }
+
+            // Someone submitted a smart contract transaction.
+            ExecuteContractTransaction(context, transaction);
+
+            base.UpdateUTXOSet(context, transaction);
+
+            this.Logger.LogTrace("(-)");
+        }
+
+        /// <inheritdoc />
+        public override void CheckMaturity(UnspentOutputs coins, int spendHeight)
```

Comment:
this is that method where if you were to remove this method, no compile fails right and no test fails?   PoS inheriting from PoW being the culprit here when PoS isn't really a specialised PoW